### PR TITLE
This class types

### LINKED ISSUE #62: 'this' typing
We need to cover `this` typing.

### PR DESCRIPTION
**Pull request by sandersn**
Add `this` types to Advanced Types.

Fixes #62

### CODE REVIEW

File: pages/Advanced Types.md [hunk 1]
```diff
@@ -313,4 +313,86 @@ As we mentioned, type aliases can act sort of like interfaces; however, there ar
...
+        .sin()
+        .add(1)
+        .currentValue();
+```
+
+Without `this` types, `ScientificCalculator` would not have been able to extend `BasicCalculator` and keep the fluent interface.
+`multiply` would have returned `BasicCalculator`, which doesn't have the `sin` method.
+However, with `this` types, `multiply` returns `this`, which is `ScientificCalculator` here.
+
+You can also use `this` types to describe mixin patterns:
```
Comment (mhegazy) [inline pages/Advanced Types.md]: not sure this is accurate.. consider:

```ts
class A {
    a(): this { return this; }
}

class B {
    b(): this { return this; }
    extend<T>(other: T): this & T { return <any>this; }
}


var b = new B();

var a = new A();

var extended = b.extend(a);

extended.a().b() // Error: b does not exist on type A
```

this types, the way we implement them, do not play well with intersection types in mixin patterns.

Comment (sandersn) [inline pages/Advanced Types.md]: Should I drop this whole mixin section? I wrote it up based on a small sample in the What's New for 1.7, but it seemed pretty undeveloped.

Comment (mhegazy) [inline pages/Advanced Types.md]: i agree. it looks like it works, untill you dot into it, and then it falls off. we should not be leading ppl there intentionally.

Comment (sandersn) [inline pages/Advanced Types.md]: Removed

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**